### PR TITLE
Fix emptyPacketMsgSize calc. plus limit max request body size and other mics. changes

### PR DESCRIPTION
**Pull request by melekes**
Refs #1762 , #1754

TODO:
- [x] update changelog

**PLEASE DO NOT SQUASH WHEN MERGING**

### CODE REVIEW

**Comment by srmo** [comment]
@melekes does it make sense for us to test this branch?

**Comment by melekes** [comment]
no, not yet.

**Comment by melekes** [comment]
@srmo can you give this branch a try now?

File: p2p/conn/connection.go [hunk 1]
```diff
@@ -788,17 +788,16 @@ func (mp PacketMsg) String() string {
 // - EOF field key + byte = 2 bytes
 // - Bytes field key = 1 bytes
 // - Uvarint length of MustMarshalBinary(bytes) = 1 or 2 bytes
-// - Struct terminator = 1 byte
-// = up to 14 bytes overhead for the packet.
+// = up to 13 bytes overhead for the packet.
 
-func emptyPacketMsgSize() int {
+func emptyPacketMsgSize(maxPayloadSize int) int {
```
Comment (melekes) [inline p2p/conn/connection.go]: still feels like a dirty hack

Comment (melekes) [reply p2p/conn/connection.go]: OK, I am going to revert this to Jae's original solution with a small tweak

File: rpc/lib/server/http_server.go [hunk 2]
```diff
@@ -202,3 +208,13 @@ func (w *ResponseWriterWrapper) WriteHeader(status int) {
 func (w *ResponseWriterWrapper) Hijack() (net.Conn, *bufio.ReadWriter, error) {
 	return w.ResponseWriter.(http.Hijacker).Hijack()
 }
+
+type maxBytesHandler struct {
+	h http.Handler
+	n int64
+}
+
+func (h maxBytesHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
Comment (alexanderbez) [inline rpc/lib/server/http_server.go]: Could we add a quick godoc -- something simple like stating that `maxBytesHandler` implements the `http.Handler interface`?

File: config/config.go [hunk 3]
```diff
@@ -299,9 +299,8 @@ type P2PConfig struct {
 	// Time to wait before flushing messages out on the connection, in ms
 	FlushThrottleTimeout int `mapstructure:"flush_throttle_timeout"`
 
-	// Maximum size of a message packet, in bytes
-	// Includes a header, which is ~13 bytes
-	MaxPacketMsgSize int `mapstructure:"max_packet_msg_size"`
+	// Maximum size of a message packet payload, in bytes
+	MaxPacketMsgPayloadSize int `mapstructure:"max_packet_msg_payload_size"`
```
Comment (ebuchman) [inline config/config.go]: We just changed this. Changing it back ?

File: consensus/state.go [hunk 4]
```diff
@@ -932,7 +932,7 @@ func (cs *ConsensusState) createProposalBlock() (block *types.Block, blockParts
 	}
 
 	// Mempool validated transactions
-	txs := cs.mempool.Reap(cs.config.MaxBlockSizeTxs)
+	txs := cs.mempool.Reap(cs.state.ConsensusParams.BlockSize.MaxTxs)
```
Comment (ebuchman) [inline consensus/state.go]: Note this means this param will be set at genesis time and can't be changed. It also has an abnormally large default

File: config/config.go (hunk 3 above)
Comment (melekes) [reply config/config.go]: yes. my bad

Jae's original code:

| -------- 1024 bytes ----------- | amino encoding 15 bytes |
limit: 1024 + 15
a) if amino changes, you need to change the `defaultAminoOverhead` const
b) 2 TM versions with different versions of amino become incompatible

My first version

| -------- 1024 bytes ----------- | (includes amino encoded header)
limit: 1024 + 15

although this version gives you control over the size of the whole package (not just payload), if you make a mistake in the calculation (like I did with emptyPacketMsgSize, it results in "maxSize is ..., but amino unmarshal 3+...." error.

My second version

| -------- 1024 bytes ----------- | amino encoding 15 bytes |
note: limit is set to 1024+15+10. 10 bytes protect us from errors above ^^ and give the opportunity to change something in amino (make it more compact / less compact)

**Comment by melekes** [comment]
> Should we try to include this in release v0.22.0 ie. #1840 ?

I think we should.

File: consensus/state.go (hunk 4 above)
Comment (melekes) [reply consensus/state.go]: > It also has an abnormally large default

set to 10000

**Pull request merged**